src/umpire/util/AllocationMap.cpp

```diff
@@ -14,249 +14,481 @@
-      const auto& pair = data->map->next();
+  RecordList();
+  RecordList(const AllocationRecord& record);
 
-      // Update key
-      data->key = pair.key;
+  ~RecordList();
 
-      if (data->map->success()) {
-        data->it = pair.value->begin();
-      }
-      else {
-        end = true;
-        data->it = data->map->begin().value->begin();
-      }
-    }
+  void push_back(AllocationRecord rec);
+  AllocationRecord pop_back();
+
+  RecordListConstIterator begin() const;
+  RecordListConstIterator end() const;
+
+  size_t size() const;
+  bool empty() const;
+  AllocationRecord* back();
+
+private:
+  Block* m_tail;
+  size_t m_length;
+};
+
+static umpire::util::FixedMallocPool list_pool(sizeof(RecordList));
+
+// Iterator for RecordList
+class RecordListConstIterator : public std::iterator<std::forward_iterator_tag, AllocationRecord>
+{
+  const RecordList *m_list;
+  RecordList::Block* m_curr;
+
+public:
+  RecordListConstIterator(const RecordList* list, bool end);
+  RecordListConstIterator(const RecordListConstIterator& other) = default;
+
+  AllocationRecord& operator*() const;
+  const AllocationRecord* operator->() const;
+  RecordListConstIterator& operator++();
+  RecordListConstIterator operator++(int);
+
+  bool operator==(const RecordListConstIterator& other);
+  bool operator!=(const RecordListConstIterator& other);
+};
+
+// Record List
+RecordList::RecordList()
+  : m_tail(nullptr), m_length(0)
+{
+}
+
+RecordList::RecordList(const AllocationRecord& record)
+  : m_tail(nullptr), m_length(0)
+{
+  push_back(record);
+  m_tail->prev = nullptr;
+}
+
+RecordList::~RecordList()
+{
+  Block* curr = m_tail;
+  while (curr) {
+    Block* prev = curr->prev;
+    block_pool.deallocate(curr);
+    curr = prev;
+  }
+}
+
+void RecordList::push_back(AllocationRecord rec)
+{
+  Block* curr = reinterpret_cast<Block*>(block_pool.allocate());
+  curr->prev = m_tail;
+  curr->rec = rec;
+  m_tail = curr;
+  m_length++;
+}
+
+AllocationRecord RecordList::pop_back()
+{
+  AllocationRecord ret{};
+  if (m_length > 0) {
+    ret = m_tail->rec;
+    Block* prev = m_tail->prev;
+    block_pool.deallocate(m_tail);
+    m_tail = prev;
+    m_length--;
   }
+  return ret;
+}
+
+RecordListConstIterator RecordList::begin() const
+{
+  return RecordListConstIterator{this, false};
+}
+
+RecordListConstIterator RecordList::end() const
+{
+  return RecordListConstIterator{this, true};
+}
+
+size_t RecordList::size() const { return m_length; }
+bool RecordList::empty() const { return size() == 0; }
+
+AllocationRecord* RecordList::back() { return &m_tail->rec; }
 
+// RecordListConstIterator
+RecordListConstIterator::RecordListConstIterator(const RecordList* list,
+                                                 bool end)
+  : m_list(list), m_curr(end ? nullptr : m_list->m_tail)
+{
+}
+
+AllocationRecord& RecordListConstIterator::operator*() const
+{
+  // WARNING: Don't call this if at end() -- dereferences a nullptr
```

Comment:
Can you please enforce this with a check?

Reply:
Done!